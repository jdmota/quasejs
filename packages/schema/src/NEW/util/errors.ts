import type { Path } from "./path";

export class SchemaError {
  readonly value: unknown;
  readonly path: Path;
  readonly message: string;
  constructor(value: unknown, path: Path, message: string) {
    this.value = value;
    this.path = path;
    this.message = message;
  }
}
