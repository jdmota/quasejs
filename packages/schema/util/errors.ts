import type { Path } from "./path";

export class SchemaError {
  readonly path: Path;
  readonly message: string;
  constructor(path: Path, message: string) {
    this.path = path;
    this.message = message;
  }
}
