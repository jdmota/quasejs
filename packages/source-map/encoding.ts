const win = typeof window !== "undefined" && window; // eslint-disable-line

const encoding = {
  encode: win ? win.btoa : (str: string) => Buffer.from(str).toString("base64"),
  decode: win
    ? win.atob
    : (str: string) => Buffer.from(str, "base64").toString(),
};

export default encoding;
