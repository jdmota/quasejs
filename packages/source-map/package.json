{
  "private": true,
  "name": "@quase/source-map",
  "version": "0.1.0",
  "description": "Source map tools",
  "license": "MIT",
  "homepage": "https://github.com/quase/quasejs",
  "repository": "https://github.com/quase/quasejs/tree/master/packages/source-map",
  "main": "dist/index.js",
  "dependencies": {
    "source-map": "^0.5.7"
  }
}
