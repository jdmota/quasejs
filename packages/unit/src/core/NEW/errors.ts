import concordance from "concordance";
import { coloredConcordanceOptions } from "./concordance-options";

export type ErrorOpts = Readonly<{
  error: unknown;
  stack: string;
  user: boolean;
  uncaught?: boolean;
  overrideStack?: boolean;
}>;

export type SimpleError = Readonly<{
  name: string;
  message: string;
  stack: string;
  diff?: string;
  user: boolean;
  uncaught: boolean;
}>;

export function processError(opts: ErrorOpts): SimpleError {
  const { error, stack, user, uncaught } = opts;
  if (error != null && typeof error === "object") {
    let diff = undefined;
    if ("diff" in error && typeof error.diff === "string") {
      diff = error.diff;
    } else if ("actual" in error && "expected" in error) {
      diff = concordance.diff(
        error.actual,
        error.expected,
        coloredConcordanceOptions
      );
    }
    const err = error as Record<string, unknown>;
    return {
      name: (err.name ?? "Error") + "",
      message: (err.message ?? "") + "",
      stack: opts.overrideStack ? stack : (err.stack ?? stack) + "",
      diff,
      user,
      uncaught: uncaught ?? false,
    };
  }
  if (typeof error === "string") {
    return {
      name: "Error",
      message: error,
      stack,
      user,
      uncaught: uncaught ?? false,
    };
  }
  return {
    name: "UnknownError",
    message: error + "",
    stack,
    user,
    uncaught: uncaught ?? false,
  };
}
