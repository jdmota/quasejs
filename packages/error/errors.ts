import stackParser from "error-stack-parser";
import { slash, prettify } from "../util/path-url";
import { type SourceMapExtractor } from "../source-map/extractor";
import type { Optional } from "../util/miscellaneous";

export const ignoreStackTraceRe =
  /StackTrace\$\$|ErrorStackParser\$\$|StackTraceGPS\$\$|StackGenerator\$\$/;
export const ignoreFileRe =
  /^([^()\s]*\/quasejs\/packages\/[^()\s/]+\/dist\/[^()\s]*|[^()\s]*\/node_modules\/@quase\/[^()\s]+|[^()\s/]+\.js|node:[^()\s]+|internal(\/[^()\s/]+)?\/[^()\s]+\.js|native)$/;

export type ErrorBeautifyOpts = Readonly<{
  extractor?: SourceMapExtractor;
  ignore?: Optional<{
    test(text: string): boolean;
  }>;
}>;

export type ErrorBeautifySimpleOpts = Readonly<{
  ignore?: Optional<{
    test(text: string): boolean;
  }>;
}>;

type StackFrame = stackParser.StackFrame & { fileName: string };

function excludeFramesWithoutFilename(
  v: stackParser.StackFrame
): v is StackFrame {
  return !!v.fileName;
}

export type BeautifiedStackLine = {
  file: string;
  code: string | null;
  name: string | null;
  line: number | null;
  column: number | null;
  args: readonly any[] | null;
};

export async function beautify(
  originalStack: string,
  options: ErrorBeautifyOpts = {}
) {
  const extractor = options && options.extractor;
  const { originalFirst, cleaned } = cleanErrorFrames(originalStack, options);
  const cleaned2 = await Promise.all(
    cleaned.map(async stackLine => {
      if (extractor && stackLine.line && stackLine.column) {
        try {
          const pos = await extractor.getOriginalLocation(stackLine.file, {
            line: stackLine.line,
            column: stackLine.column,
          });
          if ("line" in pos) {
            if (pos.originalFile != null) {
              stackLine.file = pos.originalFile;
              stackLine.code = pos.originalCode;
              stackLine.name = pos.name || stackLine.name;
              stackLine.line = pos.line;
              stackLine.column = pos.column;
            }
          } else {
            stackLine.code = pos.code;
          }
        } catch (e) {
          // Ignore
        }
      }
      return stackLine;
    })
  );
  return printCleanedErrorFrames(originalStack, {
    originalFirst,
    cleaned: cleaned2,
  });
}

export function beautifySimple(
  originalStack: string,
  options: ErrorBeautifySimpleOpts = {}
) {
  const { originalFirst, cleaned } = cleanErrorFrames(originalStack, options);
  return printCleanedErrorFrames(originalStack, { originalFirst, cleaned });
}

export function cleanErrorFrames(
  originalStack: string,
  options: ErrorBeautifySimpleOpts = {}
) {
  const ignore = options && options.ignore;
  const error = { name: "", message: "", stack: originalStack };

  const originalFrames = stackParser
    .parse(error)
    .filter(excludeFramesWithoutFilename);

  const frames = originalFrames.filter(({ fileName, functionName }) => {
    const file = slash(fileName);
    if (ignore && (ignore.test(fileName) || ignore.test(file))) {
      return false;
    }
    // Electron
    if (
      file.includes(".app/Contents/Resources/electron.asar") ||
      file.includes(".app/Contents/Resources/default_app.asar")
    ) {
      return false;
    }
    return (
      !ignoreFileRe.test(file) && !ignoreStackTraceRe.test(functionName || "")
    );
  });

  const originalFirst: StackFrame | undefined = originalFrames[0];

  if (frames.length === 0 && originalFirst) {
    frames.push(originalFirst);
  }

  return {
    originalFirst,
    cleaned: frames.map(
      ({ fileName, functionName, args, lineNumber, columnNumber }) => {
        const stackLine: BeautifiedStackLine = {
          file: fileName,
          code: null,
          name: functionName ?? null,
          line: lineNumber ?? null,
          column: columnNumber ?? null,
          args: args ?? null,
        };
        return stackLine;
      }
    ),
  };
}

function textLine(stackLine: BeautifiedStackLine) {
  return `${stackLine.name}${
    stackLine.args ? `(${stackLine.args.join(", ")})` : ""
  }`;
}

export function printCleanedErrorFrames(
  originalStack: string,
  { originalFirst, cleaned }: ReturnType<typeof cleanErrorFrames>
) {
  const cleanedText = cleaned.map(
    s => `${textLine(s)} (${prettify(s.file)}:${s.line}:${s.column})`
  );

  const title =
    originalFirst && originalFirst.source
      ? originalStack.split(originalFirst.source).shift()
      : originalStack;

  const lines = cleanedText.map(x => `    ${x}`).join("\n");
  const stack = `${title}${lines}`;

  const first = cleaned[0];

  return {
    stack,
    source: first && first.file ? first : null,
  };
}

export function getStack(offset?: number) {
  let error = new Error();

  // Not all browsers generate the `stack` property
  // Safari <=7 only, IE <=10 - 11 only
  if (!error.stack) {
    try {
      throw error;
    } catch (err: unknown) {
      error = err as Error;
    }
  }

  const stack = error.stack as string;
  const arr = stack.split("\n");
  arr.splice(1, offset != null && offset > 1 ? offset : 1);
  return arr.join("\n");
}

type Position = Readonly<{
  line?: Optional<number>;
  column?: Optional<number>;
}>;

export function positionToString(loc: Optional<Position>) {
  if (loc) {
    if (loc.line != null) {
      if (loc.column != null) {
        return `${loc.line}:${loc.column}`;
      }
      return `${loc.line}`;
    }
  }
  return "";
}

// TODO use https://nodejs.org/en/blog/release/v22.9.0#new-api-to-retrieve-execution-stack-trace

export function getCallSites(traceLimit: number = 20) {
  const target: { stack: readonly NodeJS.CallSite[] } = {} as any;

  const originalPrepare = Error.prepareStackTrace;
  const originalStackLimit = Error.stackTraceLimit;

  Error.stackTraceLimit = traceLimit;
  Error.prepareStackTrace = (error, stackTraces) => stackTraces;
  Error.captureStackTrace(target, getCallSites);

  const capturedTraces = target.stack;

  Error.prepareStackTrace = originalPrepare;
  Error.stackTraceLimit = originalStackLimit;

  return capturedTraces;
}

export function getCallSite(offset = 0) {
  return getCallSites(2 + offset)[1 + offset];
}
