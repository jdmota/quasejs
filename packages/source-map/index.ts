import { encodeBase64 } from "../util/encoding";
import { SourceMapExtractor } from "./extractor";
import { type SourceMapType, SourceMapExtractorBase } from "./extractor-base";
import {
  SourceMapConsumer,
  SourceMapGenerator,
  type Position,
} from "source-map";

export { SourceMapExtractor, SourceMapExtractorBase };
export { SourceMapConsumer, SourceMapGenerator };

// Likely that this produces a low resolution source map. Resources to consider:
// https://github.com/mozilla/source-map/issues/216
// https://github.com/Rich-Harris/sorcery
// https://github.com/aleclarson/sorcery
// https://github.com/rollup/rollup/blob/master/src/utils/collapseSourcemaps.ts
// https://github.com/babel/babel/blob/master/packages/babel-core/src/transformation/file/merge-map.js
export async function joinSourceMaps(maps: readonly SourceMapType[]) {
  if (maps.length === 0) {
    throw new Error(`No sourcemaps provided`);
  }

  if (maps.length === 1) {
    throw new Error(`Only 1 sourcemap provided`);
  }

  const consumer0 = await new SourceMapConsumer(maps[0]);
  const mergedGenerator = SourceMapGenerator.fromSourceMap(consumer0);
  consumer0.destroy();

  for (const map of maps.slice(1)) {
    const consumer = await new SourceMapConsumer(map);
    mergedGenerator.applySourceMap(consumer);
    consumer.destroy();
  }

  return mergedGenerator.toJSON();
}

export function sourceMapToString(map: SourceMapType) {
  return JSON.stringify(map);
}

export function sourceMapToUrl(map: SourceMapType) {
  return (
    "data:application/json;charset=utf-8;base64," +
    encodeBase64(sourceMapToString(map))
  );
}

// Prevent tooling from mistaking this for an actual sourceMappingURL
export const SOURCE_MAP_URL = "source" + "MappingURL";

export function sourceMapComment(url: string) {
  return `\n//# ${SOURCE_MAP_URL}=${url}`;
}

export function getOriginalLocation(
  map: SourceMapType,
  generated: Readonly<Position & { bias?: number }>
) {
  return SourceMapConsumer.with(map, null, consumer => {
    return consumer.originalPositionFor(generated);
  });
}
