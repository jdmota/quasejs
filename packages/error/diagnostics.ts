import {
  codeFrameColumns,
  type BabelCodeFrameOptions,
} from "@babel/code-frame";
import type { Optional } from "../util/miscellaneous";
import type { ReadonlyDeep } from "type-fest";
import * as colorette from "colorette";
import { emoji } from "../util/terminal";
import { prettify } from "../util/path-url";
import { beautifySimple, positionToString } from "./errors";

export type DiagnosticLoc = Readonly<{
  filePath: string | null;
  end: DiagnosticPosition | null;
  start: DiagnosticPosition;
}>;

export type DiagnosticPosition = Readonly<{
  column: number | null;
  line: number;
}>;

export type BasicDiagnostic = ReadonlyDeep<{
  category: "info" | "warn" | "error";
  message: string;
  loc?: Optional<DiagnosticLoc>;
  stack?: Optional<string>;
  codeFrame?: Optional<string>;
}>;

export type DiagnosticOpts = BasicDiagnostic &
  ReadonlyDeep<{
    related?: readonly Diagnostic[];
  }>;

export type SerializedDiagnostic = BasicDiagnostic &
  ReadonlyDeep<{
    related?: readonly SerializedDiagnostic[];
  }>;

export class Diagnostic {
  readonly category: "info" | "warn" | "error";
  readonly message: string;
  readonly loc: Optional<DiagnosticLoc>;
  readonly stack: Optional<string>;
  readonly codeFrame: Optional<string>;
  readonly related: readonly Diagnostic[];

  constructor({
    category,
    message,
    loc,
    stack,
    codeFrame,
    related,
  }: DiagnosticOpts) {
    this.category = category;
    this.message = message;
    this.loc = loc;
    this.stack = stack;
    this.codeFrame = codeFrame;
    this.related = related || [];
  }

  attachStack(func: Function) {
    if (!this.stack) {
      Error.captureStackTrace(this, func);
    }
  }
}

function indentStr(str: string, indent: string) {
  return str.replace(/^(?!\s*$)/gm, indent);
}

const formatMessage = {
  error: (message: string) =>
    `${emoji.error}  ${colorette.red(colorette.bold(message))}`,
  warn: (message: string) =>
    `${emoji.warning}  ${colorette.yellow(colorette.bold(message))}`,
  info: (message: string) =>
    `${emoji.info}  ${colorette.blue(colorette.bold(message))}`,
};

function formatDiagnosticHelper(
  diagnostic: Diagnostic,
  indent: string,
  lines: string[]
) {
  const { category, message, loc, stack, codeFrame, related } = diagnostic;

  lines.push(`${indent}${formatMessage[category](message)}`);

  if (loc) {
    lines.push(`${indent}${colorette.gray(locToString(loc))}`);
  }

  if (codeFrame) {
    lines.push(indentStr(codeFrame, indent));
  }

  if (stack) {
    lines.push(indentStr(colorette.gray(beautifySimple(stack).stack), indent));
  }

  if (related.length) {
    lines.push(`${indent}${colorette.bold("Related:")}`);

    for (const d of related) {
      formatDiagnosticHelper(d, indent + "  ", lines);
    }
  }
}

export function formatDiagnostic(diagnostic: Diagnostic) {
  const lines: string[] = [];
  formatDiagnosticHelper(diagnostic, "", lines);
  return lines.join("\n");
}

export function throwDiagnostic(diagnostic: DiagnosticOpts): never {
  throw new Diagnostic(diagnostic);
}

export function createDiagnostic(diagnostic: DiagnosticOpts): Diagnostic {
  return new Diagnostic(diagnostic);
}

export function createDiagnosticFromError(
  error: Error,
  category: Diagnostic["category"] = "error"
) {
  return createDiagnostic({
    category,
    message: `${error.name}: ${error.message}`,
    stack: error.stack
      ? error.stack.slice(error.stack.indexOf("\n") + 1)
      : undefined,
  });
}

export function createDiagnosticFromAny(
  value: unknown,
  category: Diagnostic["category"] = "error"
) {
  if (isDiagnostic(value)) {
    return value;
  }
  if (value instanceof Error) {
    return createDiagnosticFromError(value, category);
  }
  return createDiagnostic({
    category,
    message: value + "",
  });
}

export function isDiagnostic(value: unknown): value is Diagnostic {
  return value instanceof Diagnostic;
}

export function serializeDiagnostic(value: Diagnostic): SerializedDiagnostic {
  return {
    ...value,
    related: value.related && value.related.map(serializeDiagnostic),
  };
}

export function deserializeDiagnostic(value: SerializedDiagnostic): Diagnostic {
  return createDiagnostic({
    ...value,
    related: value.related && value.related.map(deserializeDiagnostic),
  });
}

export function createCodeFrame({
  code,
  loc,
  options,
}: {
  code: string;
  loc: DiagnosticLoc;
  options?: BabelCodeFrameOptions;
}) {
  return codeFrameColumns(
    code,
    {
      start: {
        line: loc.start.line,
        column: loc.start.column ?? undefined,
      },
      end: loc.end
        ? {
            line: loc.end.line,
            column: loc.end.column ?? undefined,
          }
        : undefined,
    },
    options
  );
}

function locToString(loc: Optional<DiagnosticLoc>) {
  if (loc) {
    const start = `${prettify(
      loc.filePath || "Unknown file"
    )}:${positionToString(loc.start)}`;
    if (loc.end == null) {
      return start;
    }
    return `${start}-${positionToString(loc.end)}`;
  }
  return "";
}
