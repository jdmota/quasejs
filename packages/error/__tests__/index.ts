import { expect, it } from "@jest/globals";
import { SourceMapExtractor } from "../../source-map";
import { beautify, getStack, positionToString } from "../errors";
import { createDiagnosticFromError, formatDiagnostic } from "../diagnostics";

const cwd = process.cwd();

function clean(string: string) {
  return string
    .split(cwd)
    .join("")
    .replace(/\\/g, "/")
    .split("\n")
    .filter(x => !/node_modules/.test(x))
    .join("\n");
}

function a(offset?: number) {
  return b(offset);
}

function b(offset?: number) {
  return c(offset);
}

function c(offset?: number) {
  return clean(getStack(offset));
}

it("getStack", () => {
  expect(a()).toMatchSnapshot("undefined");
  expect(a(-1)).toMatchSnapshot("-1");
  expect(a(0)).toMatchSnapshot("0");
  expect(a(1)).toMatchSnapshot("1");
  expect(a(2)).toMatchSnapshot("2");
  expect(a(3)).toMatchSnapshot("3");
});

it("beautify", async () => {
  const stack = getStack();
  const extractor = new SourceMapExtractor();

  expect(
    (await beautify(stack, { extractor, ignore: /node_modules/ })).stack
  ).toMatchSnapshot();
});

it("beautify with title", async () => {
  const stack = new Error("title").stack ?? "";
  const extractor = new SourceMapExtractor();

  expect(
    (await beautify(stack, { extractor, ignore: /node_modules/ })).stack
  ).toMatchSnapshot();
});

it("handle multine error message correctly", async () => {
  const stack = new Error("multine\nerror\nmessage").stack ?? "";

  expect(
    (await beautify(stack, { ignore: /node_modules/ })).stack
  ).toMatchSnapshot();
});

it("handle just the title fine", async () => {
  const stack = "multine\nerror\nmessage";

  expect(
    (await beautify(stack, { ignore: /node_modules/ })).stack
  ).toMatchSnapshot();
});

it("keep at least one stack line", async () => {
  const stack = getStack(2);

  expect(
    (await beautify(stack, { ignore: /node_modules/ })).stack
  ).toMatchSnapshot();
});

it("positionToString", () => {
  expect(positionToString({})).toEqual("");
  expect(positionToString({ line: 1 })).toEqual("1");
  expect(positionToString({ column: 2 })).toEqual("");
  expect(positionToString({ line: 1, column: 2 })).toEqual("1:2");
});

it("diagnostics", () => {
  const d = createDiagnosticFromError(new Error("message"));
  expect(formatDiagnostic(d)).toMatchSnapshot();
});
