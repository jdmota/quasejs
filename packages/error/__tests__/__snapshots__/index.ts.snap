// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`beautify 1`] = `
"Error: 
    Object.getStack (packages/error/__tests__/index.ts:39:17)"
`;

exports[`beautify with title 1`] = `
"Error: title
    Object.<anonymous> (packages/error/__tests__/index.ts:48:17)"
`;

exports[`getStack: -1 1`] = `
"Error: 
    at getStack (/packages/error/__tests__/index.ts:26:16)
    at c (/packages/error/__tests__/index.ts:22:10)
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:31:10)
    at new Promise (<anonymous>)"
`;

exports[`getStack: 0 1`] = `
"Error: 
    at getStack (/packages/error/__tests__/index.ts:26:16)
    at c (/packages/error/__tests__/index.ts:22:10)
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:32:10)
    at new Promise (<anonymous>)"
`;

exports[`getStack: 1 1`] = `
"Error: 
    at getStack (/packages/error/__tests__/index.ts:26:16)
    at c (/packages/error/__tests__/index.ts:22:10)
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:33:10)
    at new Promise (<anonymous>)"
`;

exports[`getStack: 2 1`] = `
"Error: 
    at c (/packages/error/__tests__/index.ts:22:10)
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:34:10)
    at new Promise (<anonymous>)"
`;

exports[`getStack: 3 1`] = `
"Error: 
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:35:10)
    at new Promise (<anonymous>)"
`;

exports[`getStack: undefined 1`] = `
"Error: 
    at getStack (/packages/error/__tests__/index.ts:26:16)
    at c (/packages/error/__tests__/index.ts:22:10)
    at b (/packages/error/__tests__/index.ts:18:10)
    at Object.a (/packages/error/__tests__/index.ts:30:10)
    at new Promise (<anonymous>)"
`;

exports[`handle just the title fine 1`] = `
"multine
error
message"
`;

exports[`handle multine error message correctly 1`] = `
"Error: multine
error
message
    Object.<anonymous> (packages/error/__tests__/index.ts:57:17)"
`;

exports[`keep at least one stack line 1`] = `
"Error: 
    Promise.finally.completed (node_modules/jest-circus/build/jestadapterinit.js:1559:28)"
`;
